pinot-plugins/pinot-stream-ingestion/pinot-pulsar/src/main/java/org/apache/pinot/plugin/stream/pulsar/PulsarStreamMessage.java

```diff
@@ -0,0 +1,47 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.plugin.stream.pulsar;
+
+import javax.annotation.Nullable;
+import org.apache.pinot.spi.stream.StreamMessage;
+import org.apache.pulsar.client.api.MessageId;
+
+public class PulsarStreamMessage extends StreamMessage<byte[]> {
+
+    private final MessageId _messageId;
```

Comment:
is this member variable needed when the same messageId is also added as metadata ?

Reply:
The MessageId is a "complex" Java object, and as such a serialized version of it is added to the metadata, as a representational string and as a Base64 encoded byte array.  It seemed worth the tradeoff to maintain a pointer to the original object rather than doing a round trip parse from a Base64 encoded byte array stored in the metadata map.

Reply:
If it is already added to metadata, I would avoid the additional reference that is maintained in this class. It is unlikely to be used by realtime engine itself. So, better not to expose something unless it is really needed.

Reply:
> Is the calling of getNextStreamPartitionMsgOffsetAtIndex infrequent enough that the Base64 round trip would be worth the benefit?

`getNextStreamPartitionMsgOffsetAtIndex` is called while looping through every message in the batch. 

> if we keep the reference in PulsarStreamMessage I can move the check in PulsarMetadataExtractor so that we are only storing the MessageId in metadata when storeMetadata is true, saving some cycles in the case when metadata extract is not enabled.

Yeah. I think you should extract the `MessageId` metadata only when `storeMetadata` is true. This change should not impact existing pulsar consumers that do not enable record metadata extraction. That's my primary concern.